Mathlib/Algebra/Order/Incidence.lean

```diff
@@ -0,0 +1,645 @@
+
+section Ring
+variable (𝕜) [Ring 𝕜] [Preorder α] [Preorder β]
+
+section DecidableLe
+variable [DecidableRel ((· ≤ ·) : α → α → Prop)] [DecidableRel ((· ≤ ·) : β → β → Prop)]
+
+lemma zeta_prod_apply (a b : α × β) : zeta 𝕜 a b = zeta 𝕜 a.1 b.1 * zeta 𝕜 a.2 b.2 := by
+  simp [← ite_and, Prod.le_def, and_comm]
+
+lemma zeta_prod_mk (a₁ a₂ : α) (b₁ b₂ : β) :
+    zeta 𝕜 (a₁, b₁) (a₂, b₂) = zeta 𝕜 a₁ a₂ * zeta 𝕜 b₁ b₂ := zeta_prod_apply _ _ _
+
+end DecidableLe
+
+variable {𝕜} (f f₁ f₂ : IncidenceAlgebra 𝕜 α) (g g₁ g₂ : IncidenceAlgebra 𝕜 β)
+
+/-- The cartesian product of two incidence algebras. -/
+protected def prod : IncidenceAlgebra 𝕜 (α × β) where
+  toFun x y := f x.1 y.1 * g x.2 y.2
+  eq_zero_of_not_le' x y hxy := by
+    rw [Prod.le_def, not_and_or] at hxy
+    obtain hxy | hxy := hxy <;> simp [apply_eq_zero_of_not_le hxy]
+
+lemma prod_mk (a₁ a₂ : α) (b₁ b₂ : β) : f.prod g (a₁, b₁) (a₂, b₂) = f a₁ a₂ * g b₁ b₂ := rfl
+@[simp] lemma prod_apply (x y : α × β) : f.prod g x y = f x.1 y.1 * g x.2 y.2 := rfl
+
+/-- This is a version of `IncidenceAlgebra.prod_mul_prod` that works over non-commutative rings. -/
+lemma prod_mul_prod' [LocallyFiniteOrder α] [LocallyFiniteOrder β] [@DecidableRel (α × β) (· ≤ ·)]
+    (h : ∀ a₁ a₂ a₃ b₁ b₂ b₃,
+        f₁ a₁ a₂ * g₁ b₁ b₂ * (f₂ a₂ a₃ * g₂ b₂ b₃) = f₁ a₁ a₂ * f₂ a₂ a₃ * (g₁ b₁ b₂ * g₂ b₂ b₃)) :
+    f₁.prod g₁ * f₂.prod g₂ = (f₁ * f₂).prod (g₁ * g₂) := by
+  ext x y; simp [Icc_prod_def, sum_mul_sum, h, sum_product]
+
+@[simp]
+lemma one_prod_one [DecidableEq α] [DecidableEq β] :
+    (1 : IncidenceAlgebra 𝕜 α).prod (1 : IncidenceAlgebra 𝕜 β) = 1 := by
+  ext x y; simp [Prod.ext_iff, ← ite_and, and_comm]
+
+@[simp]
+lemma zeta_prod_zeta [@DecidableRel α (· ≤ ·)] [@DecidableRel β (· ≤ ·)] :
+    (zeta 𝕜).prod (zeta 𝕜) = (zeta 𝕜 : IncidenceAlgebra 𝕜 (α × β)) := by
+  ext x y hxy; simp [hxy, hxy.1, hxy.2]
+
+end Ring
+
+section CommRing
+variable [CommRing 𝕜] [Preorder α] [Preorder β] [LocallyFiniteOrder α] [LocallyFiniteOrder β]
+  [@DecidableRel (α × β) (· ≤ ·)] (f₁ f₂ : IncidenceAlgebra 𝕜 α) (g₁ g₂ : IncidenceAlgebra 𝕜 β)
+
+@[simp]
+lemma prod_mul_prod : f₁.prod g₁ * f₂.prod g₂ = (f₁ * f₂).prod (g₁ * g₂) :=
+  prod_mul_prod' _ _ _ _ fun _ _ _ _ _ _ ↦ mul_mul_mul_comm ..
+
+end CommRing
+end Preorder
+
+section PartialOrder
+variable (𝕜) [Ring 𝕜] [PartialOrder α] [PartialOrder β] [LocallyFiniteOrder α]
+  [LocallyFiniteOrder β] [DecidableEq α] [DecidableEq β] [DecidableRel ((· ≤ ·) : α → α → Prop)]
+  [DecidableRel ((· ≤ ·) : β → β → Prop)]
+
+/-- The Möbius function on a product order. Based on lemma 2.1.13 of Incidence Algebras by Spiegel
+and O'Donnell. -/
+@[simp]
+lemma mu_prod_mu : (mu 𝕜).prod (mu 𝕜) = (mu 𝕜 : IncidenceAlgebra 𝕜 (α × β)) := by
+  refine left_inv_eq_right_inv ?_ zeta_mul_mu
+  rw [← zeta_prod_zeta, prod_mul_prod', mu_mul_zeta, mu_mul_zeta, one_prod_one]
+  exact fun _ _ _ _ _ _ ↦ Commute.mul_mul_mul_comm (by simp : _ = _) _ _
+
+end PartialOrder
+end Prod
+
+section Euler
```

Comment:
This one looks lonely and has no API, I'd say this should wait until a later PR when you have something to do with it!

Reply:
Yeah, fair